R/waterfall_align.R

```diff
@@ -31,81 +31,127 @@ waterfall_align <- function(p2, p1, p3, p4, p5,
     }
 
     # define the ggplot's as grobs and create a blank plot
-    gA <- suppressWarnings(ggplot2::ggplotGrob(p2))
-    gB <- ggplot2::ggplotGrob(p1)
-    gC <- ggplot2::ggplotGrob(p3)
-    if (!is.null(p5)) {
-        gE <- ggplot2::ggplotGrob(p5)
-    }
+    genes_grob <- suppressWarnings(ggplot2::ggplotGrob(genes))
+
+    heatmap_grob <- ggplot2::ggplotGrob(heatmap)
+    
+    ## Strip out legends and plot separately
+    ## https://github.com/baptiste/gridextra/wiki/arranging-ggplot#legends
+    ind_legend <- grep("guide", heatmap_grob$layout$name)
+    heatmap_legend <- heatmap_grob[["grobs"]][[ind_legend]]
+    heatmap_width <- sum(heatmap_legend$width)
+    heatmap_grob <- ggplot2::ggplotGrob(heatmap + theme(legend.position="none"))
```

Comment:
Moving the legends to a separate column of grobs means that they will never overlap the plots or spill over the side, and likewise will only take up as much room as they need